File: online-judges-and-contests/codewars/2kyu-6by6/6by6.go
package kata

import (
	"sync"
)

type ViewClue struct {
	Front int
	Back  int
}

type BuildingRow []int

const gridSize = 6

type PuzzleSolver struct {
	candidateCols [][]BuildingRow
	candidateRows [][]BuildingRow
	columnClues   []ViewClue
	rowClues      []ViewClue
	allPerms      []BuildingRow
	permsMap      map[string]bool
}

func nSolvePuzzle(clues []int) *PuzzleSolver {
	columnClues, rowClues := transformClues(clues)
	perms := generatePermutations(generateFloorOptions(gridSize))

	return &PuzzleSolver{
		allPerms:    perms,
		columnClues: columnClues,
		rowClues:    rowClues,
		permsMap:    createPermsMap(perms),
	}
}

func (ps *PuzzleSolver) Solve() [][]int {
	var wg sync.WaitGroup
	ps.candidateRows = ps.filterRows(ps.rowClues, ps.allPerms, &wg)
	ps.candidateCols = ps.filterRows(ps.columnClues, ps.allPerms, &wg)
	wg.Wait()
	return ps.assembleGrid()
}

func (ps *PuzzleSolver) assembleGrid() [][]int {
	for !ps.isUniqueSolution() {
		var wg sync.WaitGroup
		wg.Add(len(ps.candidateRows) + len(ps.candidateCols))

		for i := range ps.candidateRows {
			go func(i int) {
				defer wg.Done()
				ps.candidateRows[i] = filterValidRows(ps.candidateRows[i], ps.candidateCols, i)
			}(i)
		}

		for i := range ps.candidateCols {
			go func(i int) {
				defer wg.Done()
				ps.candidateCols[i] = filterValidRows(ps.candidateCols[i], ps.candidateRows, i)
			}(i)
		}

		wg.Wait()
	}

	result := make([][]int, gridSize)
	for i, rows := range ps.candidateRows {
		result[i] = rows[0]
	}
	return result
}

func filterValidRows(rows []BuildingRow, opposingRows [][]BuildingRow, position int) (result []BuildingRow) {
	for _, row := range rows {
		if isRowValid(row, position, opposingRows) {
			result = append(result, row)
		}
	}
	return result
}

func isRowValid(row BuildingRow, position int, opposingRows [][]BuildingRow) bool {
	for i, value := range row {
		if !containsValueAtPosition(value, position, opposingRows[i]) {
			return false
		}
	}
	return true
}

func containsValueAtPosition(value, position int, rows []BuildingRow) bool {
	for _, row := range rows {
		if row[position] == value {
			return true
		}
	}
	return false
}

func (ps *PuzzleSolver) isUniqueSolution() bool {
	for _, cols := range ps.candidateCols {
		if len(cols) > 1 {
			return false
		}
	}
	for _, rows := range ps.candidateRows {
		if len(rows) > 1 {
			return false
		}
	}
	return true
}

func (ps *PuzzleSolver) filterRows(clues []ViewClue, rows []BuildingRow, wg *sync.WaitGroup) [][]BuildingRow {
	result := make([][]BuildingRow, len(clues))
	for i, clue := range clues {
		wg.Add(1)
		go func(i int, clue ViewClue) {
			defer wg.Done()
			var validRows []BuildingRow
			for _, row := range rows {
				if row.matchesClue(clue) {
					validRows = append(validRows, row)
				}
			}
			result[i] = validRows
		}(i, clue)
	}
	return result
}

func transformClues(clues []int) (columnClues, rowClues []ViewClue) {
	for i := 0; i < len(clues)/2; i++ {
		c := ViewClue{
			Front: clues[i],
			Back:  clues[gridSize*(3+i/gridSize)-(i%gridSize+1)],
		}
		if i < gridSize {
			columnClues = append(columnClues, c)
		} else {
			c.Front, c.Back = c.Back, c.Front
			rowClues = append(rowClues, c)
		}
	}
	return columnClues, rowClues
}

func generatePermutations(source []int) (result []BuildingRow) {
	p := make([]int, len(source))
	for p[0] < len(p) {
		result = append(result, createPermutation(source, p))
		advancePermutation(p)
	}
	return result
}

func createPermsMap(perms []BuildingRow) map[string]bool {
	permsMap := make(map[string]bool)
	for _, perm := range perms {
		permsMap[rowKey(perm)] = true
	}
	return permsMap
}

func rowKey(row BuildingRow) string {
	key := ""
	for _, val := range row {
		key += string(rune(val + '0'))
	}
	return key
}

func advancePermutation(p []int) {
	for i := len(p) - 1; i >= 0; i-- {
		if i == 0 || p[i] < len(p)-i-1 {
			p[i]++
			return
		}
		p[i] = 0
	}
}

func createPermutation(source, p []int) BuildingRow {
	result := make(BuildingRow, len(source))
	copy(result, source)
	for i, v := range p {
		result[i], result[i+v] = result[i+v], result[i]
	}
	return result
}

func (br BuildingRow) countVisibleBuildings() (count int) {
	max := 0
	for _, val := range br {
		if val > max {
			count++
			max = val
		}
	}
	return count
}

func generateFloorOptions(size int) []int {
	result := make([]int, size)
	for i := 0; i < size; i++ {
		result[i] = i + 1
	}
	return result
}

func (br BuildingRow) matchesClue(clue ViewClue) bool {
	return (clue.Front == 0 || br.countVisibleBuildings() == clue.Front) &&
		(clue.Back == 0 || br.reverseRow().countVisibleBuildings() == clue.Back)
}

func (br BuildingRow) reverseRow() BuildingRow {
	size := len(br)
	result := make(BuildingRow, size)
	for i, v := range br {
		result[size-1-i] = v
	}
	return result
}

///////////////////////////////////////////
//////////////////////////

func SolvePuzzle(clues []int) [][]int {
	return nSolvePuzzle(clues).Solve()
}

File: online-judges-and-contests/codewars/2kyu-6by6/6by6.py
from itertools import permutations
from typing import List, Tuple

class ViewClue:
    def __init__(self, front: int, back: int):
        self.front = front
        self.back = back

class BuildingRow(list):
    pass

grid_size = 6

class PuzzleSolver:
    def __init__(self, clues: List[int]):
        self.column_clues, self.row_clues = self.transform_clues(clues)
        self.all_perms = list(permutations(range(1, grid_size + 1)))
        self.perms_map = {tuple(perm): True for perm in self.all_perms}

    def solve(self) -> List[List[int]]:
        self.candidate_rows = self.filter_rows(self.row_clues, self.all_perms)
        self.candidate_cols = self.filter_rows(self.column_clues, self.all_perms)

        while not self.is_unique_solution():
            self.candidate_rows = [
                self.filter_valid_rows(rows, self.candidate_cols, i)
                for i, rows in enumerate(self.candidate_rows)
            ]
            self.candidate_cols = [
                self.filter_valid_rows(cols, self.candidate_rows, i)
                for i, cols in enumerate(self.candidate_cols)
            ]

        return [list(rows[0]) for rows in self.candidate_rows]

    def filter_valid_rows(self, rows: List[BuildingRow], opposing_rows: List[List[BuildingRow]], position: int) -> List[BuildingRow]:
        return [row for row in rows if self.is_row_valid(row, position, opposing_rows)]

    def is_row_valid(self, row: BuildingRow, position: int, opposing_rows: List[List[BuildingRow]]) -> bool:
        for i, value in enumerate(row):
            if not self.contains_value_at_position(value, position, opposing_rows[i]):
                return False
        return True

    def contains_value_at_position(self, value: int, position: int, rows: List[BuildingRow]) -> bool:
        return any(row[position] == value for row in rows)

    def is_unique_solution(self) -> bool:
        return all(len(rows) == 1 for rows in self.candidate_rows) and all(len(cols) == 1 for cols in self.candidate_cols)

    def filter_rows(self, clues: List[ViewClue], rows: List[BuildingRow]) -> List[List[BuildingRow]]:
        return [
            [row for row in rows if self.matches_clue(row, clue)]
            for clue in clues
        ]

    def transform_clues(self, clues: List[int]) -> Tuple[List[ViewClue], List[ViewClue]]:
        column_clues = [
            ViewClue(clues[i], clues[grid_size * 3 - (i % grid_size) - 1])
            for i in range(grid_size)
        ]
        row_clues = [
            ViewClue(clues[grid_size * 2 + i], clues[grid_size + i])
            for i in range(grid_size)
        ]
        return column_clues, row_clues

    def matches_clue(self, row: BuildingRow, clue: ViewClue) -> bool:
        return (clue.front == 0 or self.count_visible_buildings(row) == clue.front) and (
            clue.back == 0 or self.count_visible_buildings(row[::-1]) == clue.back
        )

    def count_visible_buildings(self, row: BuildingRow) -> int:
        max_height = 0
        visible_count = 0
        for height in row:
            if height > max_height:
                visible_count += 1
                max_height = height
        return visible_count

def solve_puzzle(clues: List[int]) -> List[List[int]]:
    solver = PuzzleSolver(clues)
    return solver.solve()
